src/pr-info.ts

```diff
@@ -404,21 +412,48 @@ configSuspicious["tsconfig.json"] = makeJsonCheckerFromCore(
 // to it, ignoring some keys (JSON Patch paths).  The ignored properties are in most cases checked
 // elsewhere (dtslint), and in some cases they are irrelevant.
 function makeJsonCheckerFromCore(requiredForm: any, ignoredKeys: string[]) {
-    const diffFromReq = (text: string) => {
-        let json: any;
-        try { json = JSON.parse(text); } catch (e) { return "couldn't parse json"; }
-        jsonDiff.applyPatch(json, ignoredKeys.map(path => ({ op: "remove", path })));
-        try { return jsonDiff.compare(requiredForm, json); } catch (e) { return "couldn't diff json" };
-    };
     return (contents: string, oldText?: string) => {
-        const newDiff = diffFromReq(contents);
-        if (typeof newDiff === "string") return newDiff;
+        const diffFromReq = (json: any) => {
+            jsonDiff.applyPatch(json, ignoredKeys.map(path => ({ op: "remove", path })));
+            return jsonDiff.compare(requiredForm, json);
+        };
+        let newJson;
+        try { newJson = JSON.parse(contents); } catch (e) { return { suspect: "couldn't parse json" }; }
+        const suggestion = jsonDiff.deepClone(newJson);
+        const newDiff = diffFromReq(newJson);
         if (newDiff.length === 0) return undefined;
-        if (!oldText) return "not the required form";
-        const oldDiff = diffFromReq(oldText);
-        if (typeof oldDiff === "string") return oldDiff;
-        if (jsonDiff.compare(oldDiff, newDiff).every(({ op }) => op === "remove")) return undefined;
-        return "not the required form and not moving towards it";
+        const towardsIt = jsonDiff.deepClone(requiredForm);
+        if (oldText) {
+            let oldJson;
+            try { oldJson = JSON.parse(oldText); } catch (e) { return { suspect: "couldn't parse json" }; }
+            const oldDiff = diffFromReq(oldJson);
+            const notRemove = jsonDiff.compare(oldDiff, newDiff).filter(({ op }) => op !== "remove");
+            if (notRemove.length === 0) return undefined;
+            jsonDiff.applyPatch(newDiff, notRemove.map(({ path }) => ({ op: "remove", path })));
+            jsonDiff.applyPatch(towardsIt, newDiff.filter(({ op }) => op));
+        }
+        jsonDiff.applyPatch(suggestion, jsonDiff.compare(newJson, towardsIt));
+        // Suggest the different lines to the author
+        const suggestionLines = (
+            Object.keys(suggestion).length <= 1
+                ? prettier.format(JSON.stringify(suggestion), { tabWidth: 4, filepath: ".json" })
+                : JSON.stringify(suggestion, undefined, 4) + "\n"
+        ).split(/^/m);
+        const lines = contents.split(/^/m);
+        let i = 0;
+        while (suggestionLines[i].trim() === lines[i].trim()) i++;
+        let j = 1;
+        while (suggestionLines[suggestionLines.length - j].trim() === lines[lines.length - j].trim()) j++;
```

Comment:
Something's fishy here.  In the example that you added, there is one difference of dropping a new `await-promise` key, so the resulting suggestion has zero length.  It *looks* like GH will properly take it to show a suggestion to remove the lines but I'm not sure about it.  If it will always do that, then there should be a comment saying that this could happen and that's as intended.

Also, I'm worried about this code depending on a difference, otherwise it would find a bogus `i`/`j`.

Also, this whole code is very verbose for something to include in a small function, so I'd rather see it move out into some `makeJsonSuggestion` thing that will do all of the necessary work.

Also, I think that the code itself could be improved.  (Something along the lines of popping lines from both while they're equal, then finding the first index where they're different.)

Reply:
Done :heavy_check_mark:

Reply:
Looks much more readable -- one slight comment (feel free to ignore and I'll probably do it later: I find it even easier to read if all the variables are defined on one line:

    const lines = newText.split(/^/m), startLine = 1, endLine = lines.length;

(because this way I don't need to look into the code and figure out that `lines` doesn't change).

Reply:
No objection but I'm not sure what's best here in particular, because `startLine` and `endLine` aren't constant?